File: src/main/java/top/chenray/antilitematica/util/BedrockPlayerDetector.java
package top.chenray.antilitematica.util;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

/**
 * Detects Bedrock Edition players connected through Geyser.
 * Uses Floodgate API if available, falls back to name prefix detection.
 */
public final class BedrockPlayerDetector {

   private static Boolean floodgateChecked;
   private static Class<?> floodgateApiClass;

   private BedrockPlayerDetector() {}

   /**
    * Check if a player is a Bedrock Edition player (connected via Geyser).
    */
   public static boolean isBedrockPlayer(Player player) {
      if (player == null) return false;
      return hasFloodgateApi(player) || hasGeyserPrefix(player);
   }

   /**
    * Check using Floodgate API (most reliable).
    */
   private static boolean hasFloodgateApi(Player player) {
      try {
         if (floodgateChecked == null) {
            floodgateApiClass = Class.forName("org.geysermc.floodgate.api.FloodgateApi");
            floodgateChecked = true;
         }
         if (floodgateApiClass != null) {
            Object instance = floodgateApiClass.getMethod("getInstance").invoke(null);
            return (boolean) instance.getClass()
                  .getMethod("isFloodgatePlayer", java.util.UUID.class)
                  .invoke(instance, player.getUniqueId());
         }
      } catch (Exception ignored) {
         floodgateChecked = false;
         floodgateApiClass = null;
      }
      return false;
   }

   /**
    * Fallback: check Geyser's default name prefix (configurable in Geyser's config.yml).
    * Default prefix is "." — Bedrock players appear as ".playerName".
    */
   private static boolean hasGeyserPrefix(Player player) {
      return player.getName() != null && player.getName().startsWith(".");
   }
}

File: src/main/kotlin/icu/epochcraft/antilitematica/util/BedrockPlayerDetector.kt
package icu.epochcraft.antilitematica.util

import org.bukkit.entity.Player
import java.util.UUID

/**
 * 基岩版玩家检测（Geyser / Floodgate）。
 *
 * 基岩版客户端不存在 Litematica / Schematica 等投影 mod，
 * 检测到基岩版玩家时全部豁免，从源头消除误报。
 *
 * 检测方式：
 *   1. Floodgate API（最可靠，反射调用零依赖）
 *   2. Geyser 默认玩家名前缀 `.`（兜底）
 *
 * @author 阿清
 */
object BedrockPlayerDetector {

    @Volatile
    private var floodgateChecked: Boolean? = null
    private var floodgateApiClass: Class<*>? = null

    /** 玩家是否为基岩版玩家（Geyser 接入） */
    fun isBedrockPlayer(player: Player?): Boolean {
        if (player == null) return false
        return hasFloodgateApi(player) || hasGeyserPrefix(player)
    }

    /** Floodgate API 反射检测 */
    private fun hasFloodgateApi(player: Player): Boolean {
        try {
            if (floodgateChecked == null) {
                floodgateApiClass = Class.forName("org.geysermc.floodgate.api.FloodgateApi")
                floodgateChecked = true
            }
            val apiClass = floodgateApiClass
            if (apiClass != null) {
                val instance = apiClass.getMethod("getInstance").invoke(null)
                return instance.javaClass.getMethod("isFloodgatePlayer", UUID::class.java)
                    .invoke(instance, player.uniqueId) as Boolean
            }
        } catch (e: Exception) {
            floodgateChecked = false
            floodgateApiClass = null
        }
        return false
    }

    /** 兜底：Geyser 默认玩家名前缀 `.`（可在 Geyser 配置中修改） */
    private fun hasGeyserPrefix(player: Player): Boolean =
        player.name?.startsWith(".") == true
}
